polkadot/xcm/xcm-simulator/example/src/relay_chain/xcm_config/mod.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+pub mod asset_transactor;
+pub mod barrier;
+pub mod constants;
+pub mod location_converter;
+pub mod origin_converter;
+pub mod weigher;
+
+use crate::relay_chain::{RuntimeCall, XcmPallet};
+use frame_support::traits::{Everything, Nothing};
+use xcm_builder::{EnsureDecodableXcm, FixedRateOfFungible, FrameTransactionalProcessor};
+use xcm_executor::Config;
+
+// Generated from `decl_test_network!`
+pub type XcmRouter = EnsureDecodableXcm<crate::RelayChainXcmRouter>;
```

Comment:
very nit, what about extracting `XcmRouter` to the standalone `routing.rs`? E.g. barrier is also one-liner.

I think, also it makes sense putting a barrier and routing stuff together. The barrier validates what was routed/sent by some `SendXcm` impl, e.g. good example is `WithUniqueTopic` vs `TrailingSetTopicAsId`

Reply:
Thanks for the suggestion! I am totally open to the idea, but let me provide some larger context.

I would like to use this refactor for part of a tutorial which teaches readers how to configure XCM for various situations.

Barrier I expect will be a spot which could be configured in multiple ways, which is why it is extracted into one file even if it is a one liner.

I don't know much about XCM router. Are there other sensible configurations for it? Or usually it is just plugged into the same pallet?

Reply:
The router is usually one of the available transport options, not much configuration there.

Reply:
> Thanks for the suggestion! I am totally open to the idea, but let me provide some larger context.
> 
> Barrier I expect will be a spot which could be configured in multiple ways, which is why it is extracted into one file even if it is a one liner. That file will be used to explore the various possibilities for barrier.
> 

Ok, thank you for a context, so I agree mixing a barrier and router is not a good idea for this kind of tutorial.

> I would like to use this refactor for part of a tutorial which teaches readers how to configure XCM for various situations.
> 

> I don't know much about XCM router. Are there other sensible configurations for it? Or usually it is just plugged into the same pallet?

E.g. we have a quite complex XcmRouter for AssetHubKusama: https://github.com/polkadot-fellows/runtimes/blob/main/system-parachains/asset-hubs/asset-hub-kusama/src/xcm_config.rs#L619-L643, simplified:
```
pub XcmRouter = (
    ParentAsUmp,
    XcmpQueue,
    ToPolkadotXcmRouter,
    ToEthereumXcmRouter
);
```
I can imagine a standalone chapter for XCM routing explaining `Location`,  `SendXcm`, `ExportMessage` or creating custom implementation/feature like `WithUniqueTopic` ... so yes, for that chapter, I would extract routers to standalone files, e.g.:

**example/src/parachain/xcm_config/routing.rs**:
```
pub type XcmRouter = EnsureDecodableXcm<
       // XCMP - `(1, [$crate::Parachain(id)])` part of ParachainXcmRouter
       crate::ToSiblingParachainXcmRouter<MsgQueue>,   
       // UMP - `(1, [])` part of ParachainXcmRouter  
       crate::ToRelayChainXcmRouter<MsgQueue>, 
>;
```

**example/src/relay_chain/xcm_config/routing.rs**:
```
pub type XcmRouter = EnsureDecodableXcm<
       // DMP - `(0, [$crate::Parachain(id)])` part of RelayChainXcmRouter
       crate::ToChildParachainXcmRouter<MsgQueue>,
>;
```

Anyway, what tutorial do you mean exactly? I remember I saw your twitter post DotCodeSchool some time ago?

Reply:
@bkontur tutorial will live here, but still in progress: https://github.com/shawntabrizi/polkadot-sdk-workshop-xcm/tree/master

will be ready by the next academy.

I will spend some time to look more into XCM router config, and if it makes sense, i will push more updates to this example in Polkadot, and continue to use it as the base template for the tutorial.